src/classes/PREQ_UniqueProgramPlan_TDTM.cls

```diff
@@ -0,0 +1,74 @@
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Plan Requirement
+* @group-content ../../ApexDocContent/PlanRequirements.htm
+* @description Ensure that plan requirement always has the same program plan as its immediate parent has
+*/
+public with sharing class PREQ_UniqueProgramPlan_TDTM extends TDTM_Runnable {
+    /*******************************************************************************************************
+    * @description If unique Program Plan for plan requirements validation enabled.
+    */
+    private static boolean stgUniqueProgramPlanEnabled = UTIL_CustomSettingsFacade.getSettings().Unique_Program_Plan_Enforced__c;
+
+    /*******************************************************************************************************
+    * @description Populate an error when creating a Plan Requirement or updating a plan requirement who has
+    * different program as its immediate parent plan requirement has
+    * @param listNew the list of Plan Requirement from trigger new.
+    * @param listOld the list of Plan Requirement from trigger old.
+    * @param triggerAction which trigger event (BeforeInsert, AfterInsert, etc.).
+    * @param objResult the describe for Plan Requirement
+    * @return dmlWrapper.
+    ********************************************************************************************************/
+    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist,
+            TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+
+        if (triggerAction == TDTM_Runnable.Action.AfterInsert
+            || triggerAction == TDTM_Runnable.Action.AfterUpdate) {
+            for (Plan_Requirement__c planReq : [SELECT ID,
+                                                    Program_Plan__c,
+                                                    Plan_Requirement__c,
+                                                    Plan_Requirement__r.Program_Plan__c
+                                                FROM Plan_Requirement__c
+                                                WHERE ID in :newlist]) {
+                if (stgUniqueProgramPlanEnabled == true
+                        && planReq.Plan_Requirement__r.Program_Plan__c != null
+                        && planReq.Plan_Requirement__c != null
+                        && planReq.Program_Plan__c != planReq.Plan_Requirement__r.Program_Plan__c) {
```

Comment:
@xinnanlisf @lifewithryan I think we were talking with Vanessa this week and this logic might need to be revisited. This does not handle the use case where the Grandparent or Great-Grandparent / Great-Great-Grandparent/etc has a Program Plan but the middle children do not. In this use case, the validation would pass as a false positive. We either need to 1) traverse N levels up until the overall Program Plan is found or 2) automatically populate the Program Plan field on all child Plan Requirements or 3) ??? (something else)